guardian/atom-workshop
# Bump react to 16

## What does this change?

This PR bumps React to `16.14.0`, which will allow us to make use of React packages that use hooks (introduced in `16.8.0`) - for example the upcoming [prosemirror-editor](https://github.com/guardian/prosemirror-editor/pull/1). 

It also bumps Node from `v18` to `v20`, and bumps a few related packages to sort out and incompatibilities.

`PropTypes` is moved to its own packages from React 16, which means I've had to update the import source in lots of files. `componentWill` methods are renamed `UNSAFE_componentWill...` in React 16+ - [they're not unsafe in terms of introducing vulnerabilities](https://medium.com/@alqamabinsadiq/migrating-from-unsafe-react-lifecycle-hooks-f168f19d9550) but apparently encourage bad development practices.

`react-dates` needed to be bumped, requiring a new peer dependency ("react-with-direction"), and for `css` stylesheets to be imported directly, removing the previous `scss` files.

Changes to our linter requirements from the version bump to `eslint-plugin-react` meant a few changes to keep the linter happy, namely:
- Adding to our ESLint conf so it knows the React version
- Adding `rel="noreferrer"` to links with `target="_blank"`
- Escaping some apostrophes and double quotes

I've also removed some packages that weren't mentioned anywhere in the program, specifically:
- `es6-promise`
- `history`
- `url-search-params`
- `whatwg-fetch`

## How to test

Run the application locally using this branch, according to the instructions in the readme. Does it run as expected?

Run the tests (`yarn test`) - do they pass?
